File: src/java/com/entities/Board.java
package com.entities;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Board {
    private List<Snake> snakes;
    private List<Ladder> ladders;
    private List<Player> players;
    private Map<String, Integer> playerPositions;

    public Board() {
        snakes = new ArrayList<>();
        ladders = new ArrayList<>();
        players = new ArrayList<>();
        playerPositions = new HashMap<>();
    }

    public void addSnake(int head, int tail) {
        Snake snake = new Snake(head, tail);
        snakes.add(snake);
    }

    public void addLadder(int start, int end) {
        Ladder ladder = new Ladder(start, end);
        ladders.add(ladder);
    }

    public void addPlayer(String playerName) {
        Player player = new Player(playerName, 0);
        players.add(player);
        playerPositions.put(playerName, 0);
    }

    public List<Player> getPlayers() {
        return players;
    }

    public List<Snake> getSnakes() {return snakes;
    }

    public List<Ladder> getLadders() {
        return ladders;
    }
}

File: src/entities/Board.kt
package src.entities

import src.DEFAULT_BOARD_SIZE
import java.util.LinkedList

class Board(
    snakes: List<Snake> = LinkedList(),
    ladders: List<Ladder> = LinkedList(),
    private val size: Int = DEFAULT_BOARD_SIZE,
) {

    val dice = Dice()
    val snakeMap = hashMapOf<Int, Snake>()
    val ladderMap = hashMapOf<Int, Ladder>()

    init {
        validateBoard(snakes, ladders)

        snakes.forEach { snake ->
            snakeMap[snake.head] = snake
        }
        ladders.forEach { ladder ->
            ladderMap[ladder.start] = ladder
        }

        println("Board size $size")
    }

    private fun validateBoard(snakes: List<Snake>, ladders: List<Ladder>) {
        snakes.forEach { snake ->
            if (snake.head > size || snake.head < snake.tail)
                throw Exception("Invalid coordinates of Snake at ${snake.head} ${snake.tail}")
        }

        ladders.forEach { ladder ->
            if (ladder.end > size || ladder.end < ladder.start)
                throw Exception("Invalid coordinates of Ladder at ${ladder.start} ${ladder.end}")
        }
    }

    fun getSize(): Int {
        return size
    }
}
